core-bundle/src/EventListener/TwoFactorFrontendListener.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Exception\PageNotFoundException;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Contao\FrontendUser;
+use Contao\PageModel;
+use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorToken;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\Exception\UnauthorizedHttpException;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class TwoFactorFrontendListener
+{
+    use TargetPathTrait;
+
+    /** @var ContaoFramework */
+    private $framework;
+
+    /**
+     * @var ScopeMatcher
+     */
+    private $scopeMatcher;
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var array
+     */
+    private $supportedTokens;
+
+    public function __construct(ContaoFramework $framework, ScopeMatcher $scopeMatcher, TokenStorageInterface $tokenStorage, array $supportedTokens)
+    {
+        $this->framework = $framework;
+        $this->scopeMatcher = $scopeMatcher;
+        $this->tokenStorage = $tokenStorage;
+        $this->supportedTokens = $supportedTokens;
+    }
+
+    public function onKernelRequest(GetResponseEvent $event): void
+    {
+        // Check if is frontend request
+        if (!$this->scopeMatcher->isFrontendMasterRequest($event)) {
+            return;
+        }
+
+        $request = $event->getRequest();
+        $token = $this->tokenStorage->getToken();
+
+        // Check if is a supported token
+        if (!$token instanceof TwoFactorToken && !\in_array(\get_class($token), $this->supportedTokens, true)) {
```

Comment:
Guess that should be `||`?

Reply:
no, it can either be a `TwoFactorToken` or a supported one (`UsernamePasswordToken` by default, see https://github.com/scheb/two-factor-bundle/blob/master/DependencyInjection/Configuration.php#L89-L92)

Reply:
This is still wrong. We only need to check for `TwoFactorToken`. It's actually a `TwoFactorTokenListener` so everything else does not matter.

This configuration parameter from the two-factor-bundle means something different:
```
    # The security token classes, which trigger two-factor authentication.
    # By default the bundle only reacts to Symfony's username+password authentication. If you want to enable
    # two-factor authentication for other authentication methods, add their security token classes.
    security_tokens:
        - Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken
```
https://github.com/scheb/two-factor-bundle/blob/master/Resources/doc/configuration.md

But these possibly other tokens get wrapped into a `TwoFactorToken` anyway.